rfc-134-govuk-wide-session-cookie-and-login.md

```diff
@@ -0,0 +1,337 @@
+
+Rather than have multiple GOV.UK frontend apps worry about receiving
+the cookie and bumping its expiration time, if we manage the cookie
+entirely in VCL—parsing it into a new `GOVUK-Session-ID` header—we
+only need to worry about returning appropriate response headers.
+
+We will need to make three changes to our VCL.
+
+[fastly-cookies-practices]: https://developer.fastly.com/reference/http-headers/Cookie/#best-practices
+
+#### Preventing tampering with the cookie
+
+We can't rely on client-side cookie timeouts to log users out.  If an
+attacker gets a copy of a user's session cookie, they can include it
+in any requests they send, regardless of whether it has expired.  So
+we need to sign the cookie value and expiration time to prevent
+tampering.
+
+The cookie value will be of the form:
+
+```ruby
+"#{base64(session_id)}.#{base64(expiration_time)}.#{base64(signature)}"
+```
+
+Where the signature is computed with:
+
+```ruby
+hmac_sha256(secret_key, "#{session_id}.#{expiration_time}")
+```
+
+We will need a random string as a signing key, stored in a table:
+
+```vcl
+table env {
+  "session_key": "random string goes here"
+}
+```
+
+This approach is based on [jwt-vcl][], though we don't need a full JWT
+implementation so I have simplified it.
+
+[jwt-vcl]: https://github.com/phamann/jwt-vcl
+
+#### Changes to `vcl_recv`
+
+When receiving a request, validate the cookie: if it's ok, put the
+session ID in the `GOVUK-Session-ID` header for our apps, if it's not
+ok, unset it.
+
+```vcl
+if (req.http.Cookie ~ "govuk_session") {
+  if (req.http.Cookie:govuk_session ~ "^([a-zA-Z0-9\-_]+)?\.([a-zA-Z0-9\-_]+)?\.([a-zA-Z0-9\-_]+)?$") {
+    set req.http.GOVUK-Session-ID = digest.base64url_nopad_decode(re.group.1);
+    set req.http.GOVUK-Session-Expires = digest.base64url_nopad_decode(re.group.2);
+    set req.http.GOVUK-Session-Signature = digest.base64url_nopad_decode(re.group.3);
+    set req.http.GOVUK-Session-Expected-Signature = digest.hmac_sha256(table.lookup(env, "session_key"), req.http.GOVUK-Session-ID + "." + req.http.GOVUK-Session-Expires);
+
+    if(!digest.secure_is_equal(req.http.GOVUK-Session-Signature, req.http.GOVUK-Session-Expected-Signature) || time.is_after(now, std.integer2time(std.atoi(req.http.GOVUK-Session-Expires)))) {
+      set req.http.GOVUK-End-Session = "true";
+      unset req.http.GOVUK-Session-ID;
+    }
+
+    unset req.http.GOVUK-Session-Expires;
+    unset req.http.GOVUK-Session-Signature;
+    unset req.http.GOVUK-Session-Expected-Signature;
+  } else {
+    set req.http.GOVUK-End-Session = "true";
+  }
+}
+```
+
+#### Changes to `vcl_fetch`
+
+When fetching a response from the backend, record the new
+`GOVUK-Session-ID` value, if one is returned; and end the session, if
+the user has been logged out:
+
+```vcl
+if (beresp.http.GOVUK-Session-ID) {
+  set req.http.GOVUK-Session-ID = beresp.http.GOVUK-Session-ID;
+}
+
+if (beresp.http.GOVUK-End-Session) {
+  set req.http.GOVUK-End-Session = "true";
```

Comment:
Would apps be listening for this header to delete a session or would this be purely for VCL?

Reply:
This is purely for VCL